Flattens The Floor Plane (Camera Update Too)

LemonInTheDark opened the pull request:
## About The Pull Request

Ok so like, side map right? It makes things higher up in the world render above things lower down in the world.

Most of the time this is what we want, but it is NOT what we want for floors.
Floors are allowed to be larger then 32x32, and if they are we want them to render based off JUST their layer.
If we don't allow this grass turfs and others get cut off on their bottom edge, which looks WEIRD.

In order to make this happen, we can add TOPDOWN_LAYER to every layer on the floor plane and disable sidemap.

I've added documentation for this to VISUALS.md, and have also implemented unit test errors to prevent mixing TOPDOWN layers with non topdown planes (or vis versa).
This new test adds ~1 second to tests, which is I think a perfectly scrumpulent number.

EDIT:

I nerd sniped myself and implemented sidemap layering and lighting for cameras (also larger then 32x32 icon support for getflat)
The lighting isn't perfect, we don't handle things displaying in the void all that well (I am convinced getflat blending is broken but I have no debugger so I can't fix it properly), but it'll do.

This came up cause I had to fix another layering issue in cameras and thought I might as well go all in.
![image](https://github.com/tgstation/tgstation/assets/58055496/601b422c-f6aa-42ba-bcd9-b1faebe236e3)


## Why It's Good For The Game

Old:
![image](https://github.com/tgstation/tgstation/assets/58055496/d4102386-420d-4346-b05c-b819e62d98d0)

New:
![image](https://github.com/tgstation/tgstation/assets/58055496/1f5e303e-adee-427d-8fe3-76c8f2dbe098)


## Changelog
:cl:
fix: Grass turfs will render properly now. Reworked how floors render, please report any bugs!
fix: Cameras now properly capture lighting
fix: The layering seen in photos should better match the actual game
/:cl:

LemonInTheDark committed:
Flattens The Floor Plane

Ok so like, side map right? It makes things higher up in the world
render above things lower down in the world.

Most of the time this is what we want, but it is NOT what we want for
floors.
Floors are allowed to be larger then 32x32, and if they are we want them
to render based off JUST their layer.

In order to make this happen, we can add TOPDOWN_LAYER to every layer on
the floor plane and disable sidemap.

This resolves some issues with existing large floors (mostly grass
tiles)

I've added documentation for this to VISUALS.md, and will be implmenting
unit test errors to prevent mixing TOPDOWN layers with non topdown planes (or
vis versa)

LemonInTheDark committed:
Adds unit test, cleans up more problem children

LemonInTheDark committed:
fixes up floor lights

Kapu1178 commented:
does this also fix cult rune rendering or was that fixed a while ago? regardless this is a good change and a reminder that the byond renderer must be absolutely fucking cursed

EEASAS commented:
i kinda liked the lime coloured grass but oh well

LemonInTheDark commented:
> i kinda liked the lime coloured grass but oh well

to be clear I used the greener stuff basically on accident, the color of the grass isn't being changed at all it's just the grass tufts actually show on the south edge now

LemonInTheDark commented:
> does this also fix cult rune rendering or was that fixed a while ago? regardless this is a good change and a reminder that the byond renderer must be absolutely fucking cursed

I do not know. frankly right now it breaks vents displaying because I forgot to test for broken combos AFTER init, but that is LIFE baby

LemonInTheDark committed:
Fixes everythin else, adds proper test

I forgot to test state POST mapload. We should do that.

Also removed test in atom init since it's now redundant. keeping MAs tho

LemonInTheDark commented:
> does this also fix cult rune rendering or was that fixed a while ago? regardless this is a good change and a reminder that the byond renderer must be absolutely fucking cursed

Likely no because iirc it's not on the floor plane at all

Also this should be good now I fixed the rest of things + a unit test

LemonInTheDark committed:
OBEY ME MACHINE

LemonInTheDark committed:
begone

LemonInTheDark committed:
whoop

san7890 commented on code/modules/unit_tests/plane_sanity.dm:
```diff
@@ -0,0 +1,9 @@
+/datum/unit_test/plane_layer_sanity
+
+/datum/unit_test/plane_layer_sanity/Run()
+	// This fucker's gonna be slow, I'm sorry
```
add `priority = TEST_LONGER`?

LemonInTheDark replied:
sure

LemonInTheDark committed:
priority changes (san'd)

LemonInTheDark committed:
:newspaper2: layermanifolds

LemonInTheDark committed:
float layer use to make this sane

LemonInTheDark committed:
FUCKING MAPPERS WHY ARE YOU EDITING THE LAYER FROM 10 YEARS AGO MOTHERFUCKER

san7890 commented:
test merges going alright?

LemonInTheDark commented:
https://github.com/tgstation/tgstation/issues/84524
no :(

PapaMichae1 commented:
I don't know the specifics well enough to attempt a issue yet but cameras seem broken in this tm too

LemonInTheDark committed:
catwalk issue resolution

LemonInTheDark committed:
Handles layer for cameras properly (Basically we were equating plane and layer when they are NOT that

LemonInTheDark committed:
Resolves kudzu and also pitgrate issues

LemonInTheDark committed:
Implements a more complete version of camera lighting, roll back to this someday?

LemonInTheDark committed:
More basic camera lighting implementation (works)

LemonInTheDark committed:
Adds support for larger icons to getflat

LemonInTheDark committed:
*huffs salt*

LemonInTheDark committed:
cleanliness

LemonInTheDark committed:
removes debug code

jlsnow301 commented:
20 rounds on the tm + maintainer pr + lgtm

LemonInTheDark commented:
> ... maintainer pr ...

you should not trust me

OverwatchVoice commented:
As a result of this PR, the atmos-thermal vision is still broken, making it even more difficult to locate breaches on multi-Z maps.
